### ISSUE TITLE
Treat jQuery & Tether as dependencies in ES6 source code

### ISSUE THREAD

**Issue by vejersele** [twbs/bootstrap#17201]
First of all, congrats on the v4 release guys. Looking really nice! I'm especially excited about the new ES6 source code. However, when scanning the source code I noticed that jQuery isn't imported like the other dependencies. Because of this jQuery is also missing in the UMD definitions that are generated via grunt. This causes the UMD modules to be more or less unusable unless you include jQuery globally.

I guess the reason for this is that it's probably difficult to guess how people are going to import jQuery. Import paths and package names may differ between package managers, etc ...

However, since jQuery is the only external dependency here, I think it should be possible. For example you could ask people using requirejs to define a custom path, [which they are probably doing already](http://requirejs.org/docs/jquery.html) ... 

On npm the package for jQuery is 'jquery', so this shouldn't cause any issues. Then the only remaining problem is that babel is converting this:

``` javascript
import $ from 'jquery'
```

to 

``` javascript
factory(mod.exports, mod, global.$);
```

Which probably will work, but it is kind of risky to use '$' instead of jQuery I guess ... Maybe there is way to hook into babel's compiling step and change module names there ... 

Adding jQuery as a dependency will also enable people who are already writing ES6 code to actually import the ES6 code ...

I may be missing a lot of stuff here, so any feedback is more than welcome!

**Comment by zacechola** [comment]
FWIW Ember handles this by importing jQuery by default while allowing authors to toggle that default off.

**Comment by vejersele** [comment]
I think the simplest solution would be to just import jQuery as jQuery and then map $ to jQuery. This way you'll get a clean UMD definition, which is usable with bower or npm and requirejs, commonjs or globals ...

```javascript
import jQuery from 'jquery'

const $ = jQuery
```
assuming requirejs users create a custom path for 'jquery' ...

**Comment by hnrch02** [comment]
CC @fat

**Comment by vejersele** [comment]
Small update:

I was able to get it working just by adding import jQuery as jQuery to every JS file. If you're ok with it, I would suggest to run the unit tests directly on the UMD builds ... I see no need to compile a version without UMD definitions anymore ... 

One thing I noticed though while testing this locally, is that there are some additional issues regarding the UMD builds ... I've created a separate [issue](https://github.com/twbs/bootstrap/issues/17265) for this.

Once I fixed these 2 problems I was able to successfully run unit tests against the UMD builds. If you want I can come up with a pull request.

**Comment by krnlde** [comment]
@vejersele
```
import jQuery as $ from 'jquery'
```

**Comment by vejersele** [comment]
@krnlde That doesn't work ... 

```javascript
import {jQuery as $} from 'jquery'
```
This does work, but leaves us with the same issue as we had in the first place since it's compiling to global.jquery while we actually want global.jQuery.

**Comment by krnlde** [comment]
What about `import * as $ from jQuery`

**Comment by vejersele** [comment]
I assume you mean ```import * as $ from 'jQuery'``` which compiles to global.$ ... Also, it will use require('jQuery') in a node environment, while I don't think require() is case sensitive (at least on a OSX it isn't) it seems like a dangerous guess.

**Comment by krnlde** [comment]
Sorry for the errors, I answered from my mobile, which autocorrect, and I correctly assumed you'd guess the correct. This aside, `const $ = jQuery` doesn't look like a good practice to me.

**Comment by vejersele** [comment]
I'm aware that this isn't the cleanest solution, but at least it works ... I only see two options, either use const $ = jQuery or wrapping the component code within an IIFE with $ as a param (how it's already done right now), which is essentially the same.

**Comment by krnlde** [comment]
What about `import {$, jQuery} from 'jquery'`?

**Comment by krnlde** [comment]
I don't know enough about destructuring, but this could work also, if you can use the same key twice: `import {jQuery as $, jQuery} from 'jquery'`

**Comment by vejersele** [comment]
Not sure if I need to create another issue for this, but I also think we should specify jQuery and Tether as dependencies in the package.json and bower.json files ...

**Comment by zacechola** [comment]
@vejersele Doesn't hurt. Just x-ref the issue numbers as they're all related anyway.

**Comment by fat** [comment]
@vejersele if you havne't already, a pr would be rad

**Comment by vejersele** [comment]
Once #18313 is merged I'll put together a pull request.

**Comment by LeBenLeBen** [comment]
+1 would be cool to require jQuery.

**Comment by dbkaplun** [comment]
+100

**Comment by niftylettuce** [comment]
+1

**Comment by danieldram** [comment]
+1 (We're having issues getting Boot4 to work with an Angular 2 build, same issue jQuery undefined)

**Comment by ulrik59** [comment]
+1 ! I has to used a special plugin to inject jQuery as a global with Rollup.js. Just for Bootstrap...

**Comment by Jenso** [comment]
+1

**Comment by krnlde** [comment]
There is a "+1"-reaction now, please use this instead.

**Comment by DaSchTour** [comment]
But the +1 reaction doesn't add you as a participant ;)

**Comment by krnlde** [comment]
So you want to say bombing this thread is better than doing a second click on the subscribe button? Give that man a burrito.

**Comment by DaSchTour** [comment]
Subscribe isn't the same as participate ;)

**Comment by krnlde** [comment]
But apparently trolling is.

**Comment by DaSchTour** [comment]
You think that's a joke? Just click on the watch/unwatch button and you will see what I mean ;)

**Comment by karolyi** [comment]
linking my comment here so some attention might be thrown to it:

https://github.com/twbs/bootstrap/issues/18647#issuecomment-247882258

**Comment by mdo** [comment]
According to the comment in #20745, I think #19541 took care of this. Correct me if I'm wrong though!

**Comment by bardiharborow** [comment]
#20745 was replaced by #19541, but neither completely fix this issue. We need to work out how we are going to manage all this, especially given that native module support in Chrome isn't too far off.

**Comment by karolyi** [comment]
Well, injecting tether with webpack works this way, but when you use require('Tether') and webpack is configured correctly, it will import tether for itself.

I don't know anything about chrome native module support, but since it doesn't exist yet, you have to transpile TWBS's modules anyway, and that solves the problem.

**Comment by gijsbotje** [comment]
issue can be closed right? #22444

**Comment by karolyi** [comment]
hey, thanks for the heads up for this, changing to Popper was news to me, however I welcome it. I had problems here and there with tether, and looking forward replacing it with the next alpha release.

as for this issue, yes it should be closed.

**Comment by Johann-S** [comment]
Hmm I don't think we should close this issue because we should treat jQuery and (now) Popper.js as dependencies and it will be done here https://github.com/twbs/bootstrap/pull/22888

**Comment by Johann-S** [comment]
Closed thanks to #23735
